test/dht.spec.js

```diff
@@ -0,0 +1,651 @@
+    let daemon
+    let client
+
+    before(async function () {
+      daemon = await createDaemon(daemonOpts())
+      await daemon.start()
+    })
+
+    after(() => {
+      return daemon.stop()
+    })
+
+    afterEach(async () => {
+      await client && client.close()
+    })
+
+    it('should be able to get a value from the dth', async function () {
+      const key = '/key'
+      const value = Buffer.from('oh hello there')
+
+      client = new Client(defaultSock)
+
+      await client.attach()
+
+      try {
+        await client.dht.put(key, value)
+      } catch (err) {
+        expect(err).to.not.exist()
+      }
+
+      let data
+      try {
+        data = await client.dht.get(key)
+      } catch (err) {
+        expect(err).to.not.exist()
+      }
+
+      expect(data).to.exist()
+      expect(data).to.equalBytes(value)
+    })
+
+    it('should error if receive an invalid key', async function () {
+      client = new Client(defaultSock)
+
+      await client.attach()
+
+      try {
+        await client.dht.get(Buffer.from('/key'))
+      } catch (err) {
+        expect(err).to.exist()
+        expect(err.code).to.equal('ERR_INVALID_KEY')
+      }
+    })
+
+    it('should error if it cannot get a value', async function () {
+      client = new Client(defaultSock)
+
+      await client.attach()
+
+      try {
+        await client.dht.get('/unavailable-key')
+      } catch (err) {
+        expect(err).to.exist()
+      }
+    })
+  })
+
+  describe('findPeer', () => {
+    const sock2 = getSockPath('/tmp/p2pd-2.sock')
+    let daemonA
+    let daemonB
+    let client
+
+    before(function () {
+      return Promise.all([
+        createDaemon(daemonOpts()),
+        createDaemon(daemonOpts(sock2))
+      ]).then((res) => {
+        daemonA = res[0]
+        daemonB = res[1]
+
+        return Promise.all([
+          daemonA.start(),
+          daemonB.start()
+        ])
+      })
+    })
+
+    after(function () {
+      return Promise.all([
+        daemonA.stop(),
+        daemonB.stop()
+      ])
+    })
+
+    afterEach(async () => {
+      await client && client.close()
+    })
+
+    it('should be able to find a peer', async () => {
+      client = new Client(defaultSock)
+
+      await client.attach()
+
+      let identify
+      try {
+        identify = await client.identify()
+      } catch (err) {
+        expect(err).to.not.exist()
+      }
+
+      // close first client
+      client.close()
+
+      client = new Client(sock2)
+
+      await client.attach()
+
+      try {
+        await client.connect(identify.peerId, identify.addrs)
+      } catch (err) {
+        expect(err).to.not.exist()
+      }
+
+      let result
+      let count = 0
+      let error
+
+      do {
+        error = undefined
+        try {
+          result = await client.dht.findPeer(identify.peerId)
+        } catch (err) {
+          error = err
+        } finally {
+          count += 1
+        }
+      } while (count < 5 && error)
```

Comment:
What's the reason behind the iteration here? If this is necessary some comments around this `do...while` block would be helpful.

Reply:
Do you need the count logic in there since the timeout will take care of it in the end? Perhaps something like:

```js
let result
// Retry until we hit the test timeout since dht propagation isn't instant
while (true) {
  try {
    result = await client.dht.findPeer(identify.peerId)
    break // we've got a result, exit the loop
  } catch (err) { /* ignore errors, we only care about the test timing out */ }
}
expect(result).to.exist()
expect(result.id.toB58String()).to.equal(identify.peerId.toB58String())
```

Reply:
Seems cleaner retrying until we hit the timeout! Thanks, updated both tests that were using `promise-retry`